libvips/iofuncs/threadset.c

```diff
@@ -0,0 +1,274 @@
+                member->user_data = NULL;
+
+                /* We are free ... back on the free list!
+                 */
+                g_mutex_lock( set->lock );
+                set->free = g_slist_prepend( set->free, member );
+                g_mutex_unlock( set->lock );
+        }
+
+        /* Kill has been requested. We leave this thread on the members 
+         * list so it can be found and joined.
+         */
+
+        return( NULL );
+}
+
+/* Create a new idle member for the set.
+ */
+static VipsThreadsetMember *
+vips_threadset_add( VipsThreadset *set )
+{
+        VipsThreadsetMember *member;
+
+        member = g_new0( VipsThreadsetMember, 1 );
+        member->set = set;
+
+	vips_semaphore_init( &member->idle, 0, "idle" );
+
+	if( !(member->thread = vips_g_thread_new( "libvips worker", 
+                vips_threadset_work, member )) ) {
+                vips_semaphore_destroy( &member->idle );
+                VIPS_FREE( member );
+
+                return( NULL );
+	}
+
+        g_mutex_lock( set->lock );
+        set->members = g_slist_prepend( set->members, member );
+        set->n_threads += 1;
+        set->max_threads = VIPS_MAX( set->max_threads, set->n_threads );;
+        g_mutex_unlock( set->lock );
+
+        return( member );
+}
+
+VipsThreadset *
+vips_threadset_new( void )
+{
+        VipsThreadset *set;
+
+        set = g_new0( VipsThreadset, 1 );
+	set->lock = vips_g_mutex_new();
+
+        return( set );
+}
+
+/* Execute a task in a thread. If there are no idle threads, create a new one.
+ */
+int
+vips_threadset_run( VipsThreadset *set, 
+        const char *domain, GFunc func, gpointer data )
+{
+        VipsThreadsetMember *member;
+
+        member = NULL;
+
+        /* Try to get an idle thread.
+         */
+        g_mutex_lock( set->lock );
+        if( set->free ) {
+                member = (VipsThreadsetMember *) set->free->data;
+                set->free = g_slist_remove( set->free, member );
+        }
+        g_mutex_unlock( set->lock );
+
+        /* None? Make a new idle but not free member.
+         */
+        if( !member )
+                member = vips_threadset_add( set );
+
+        /* Still nothing? Thread create has failed.
+         */
+        if( !member )
+                return( -1 );
+
+        /* Allocate the task and set it going.
+         */
+        member->domain = domain;
+        member->func = func;
+        member->data = data;
+        member->user_data = NULL;
+        vips_semaphore_up( &member->idle );
+
+        return( 0 );
+}
+
+/* Kill a member.
+ */
+static void
+vips_threadset_kill_member( VipsThreadsetMember *member )
+{
+        VipsThreadset *set = member->set;
+
+        member->kill = TRUE;
+        vips_semaphore_up( &member->idle );
+        g_thread_join( member->thread );
+
+	vips_semaphore_destroy( &member->idle );
+
+        g_mutex_lock( set->lock );
+        set->free = g_slist_remove( set->free, member );
+        set->n_threads += 1;
```

Comment:
Since this kills a thread, perhaps it should subtract `->n_threads` instead?
```suggestion
        set->n_threads -= 1;
```

Reply:
Oooops! You're right, nice spot. Though I don't think it matters (fortunately).